node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -89,75 +89,105 @@ retrievalWorkerImpl SendActions {..} =
-        reportOrLogE "retrievalWorker: error caught " e
-        delay (30 & sec)
+        reportOrLogE "retrievalWorker mainLoopE: error caught " e
+        delay (sec 1)
         mainLoop
-    --
-    handleBlockRetrieval nodeId BlockRetrievalTask{..} =
-        handleAll (handleBlockRetrievalE nodeId brtHeader) $
+
+    -----------------
+
+    -- That's the first queue branch (task dispatching).
+    handleBlockRetrieval nodeId BlockRetrievalTask{..} = do
+        logDebug $ sformat
+            ("Block retrieval queue task received, nodeId="%build%
+             ", header="%build%", continues="%build)
+            nodeId
+            (headerHash brtHeader)
+            brtContinues
         (if brtContinues then handleContinues else handleAlternative)
             nodeId
             brtHeader
+
+    -- When we have a continuation of the chain, just try to get and apply it.
     handleContinues nodeId header = do
         logDebug $ "handleContinues: " <> pretty header
         processContHeader enqueueMsg nodeId header
+
+    -- When we have an alternative header, we should check whether it's actually
+    -- recovery mode (server side should send us headers as a proof) and then
+    -- enter recovery mode.
     handleAlternative nodeId header = do
         logDebug $ "handleAlternative: " <> pretty header
-        mhrr <- mkHeadersRequest (headerHash header)
-        case mhrr of
-            MhrrBlockAdopted ->
-                logDebug "Block already adopted, nothing to be done"
-            MhrrWithCheckpoints mgh -> do
-                logDebug "Checkpoints available, headers request assembled"
-                handleHeadersRequest nodeId header mgh
-    handleHeadersRequest nodeId header mgh = do
-        updateRecoveryHeader nodeId header
-        let cont (headers :: NewestFirst NE (BlockHeader ssc)) =
-                let oldestHeader = headers ^. _NewestFirst . _neLast
-                    newestHeader = headers ^. _NewestFirst . _neHead
-                in handleCHsValid enqueueMsg nodeId
-                                  oldestHeader (headerHash newestHeader)
-        -- If it returns w/o exception then we're:
-        --  * Either still in recovery mode
-        --  * Or just exited it and recoverVar was taken.
-        () <-
-            enqueueMsgSingle enqueueMsg (MsgRequestBlockHeaders $ Just $ S.singleton nodeId) $
-            Conversation $ requestHeaders cont mgh nodeId
-        logInfo "Recovery mode exited gracefully"
-    handleBlockRetrievalE nodeId header e = do
-        -- REPORT:ERROR 'reportOrLogW' in block retrieval worker.
+        classifyNewHeader header >>= \case
+            CHInvalid _ ->
+                logError "handleAlternative: invalid header got into retrievalWorker queue"
+            CHUseless _ ->
+                logDebug $
+                sformat ("handleAlternative: header "%build%" became useless, ignoring it")
+                        header
+            _ -> do
```

Comment:
Are we sure we want to use a wildcard here rather than being explicit? Looks like the typical fragment of code where we might want to not leave anything to chance.

Reply:
I agree.